rust/primary/src/core.rs

```diff
@@ -0,0 +1,411 @@
+                    .committee
+                    .primary(&header.author)
+                    .expect("Author of valid header is not in the committee")
+                    .primary_to_primary;
+                let bytes = bincode::serialize(&PrimaryMessage::Vote(vote))
+                    .expect("Failed to serialize our own vote");
+                let handler = self.network.send(address, Bytes::from(bytes)).await;
+                self.cancel_handlers
+                    .entry(header.round)
+                    .or_insert_with(Vec::new)
+                    .push(handler);
+            }
+        }
+        Ok(())
+    }
+
+    #[async_recursion]
+    async fn process_vote(&mut self, vote: Vote) -> DagResult<()> {
+        debug!("Processing {:?}", vote);
+
+        // Add it to the votes' aggregator and try to make a new certificate.
+        if let Some(certificate) =
+            self.votes_aggregator
+                .append(vote, &self.committee, &self.current_header)?
+        {
+            debug!("Assembled {:?}", certificate);
+
+            // Broadcast the certificate.
+            let addresses = self
+                .committee
+                .others_primaries(&self.name)
+                .iter()
+                .map(|(_, x)| x.primary_to_primary)
+                .collect();
+            let bytes = bincode::serialize(&PrimaryMessage::Certificate(certificate.clone()))
+                .expect("Failed to serialize our own certificate");
+            let handlers = self.network.broadcast(addresses, Bytes::from(bytes)).await;
+            self.cancel_handlers
+                .entry(certificate.round())
+                .or_insert_with(Vec::new)
+                .extend(handlers);
+
+            // Process the new certificate.
+            self.process_certificate(certificate)
+                .await
+                .expect("Failed to process valid certificate");
+        }
+        Ok(())
+    }
+
+    #[async_recursion]
+    async fn process_certificate(&mut self, certificate: Certificate) -> DagResult<()> {
+        debug!("Processing {:?}", certificate);
+
+        // Process the header embedded in the certificate if we haven't already voted for it (if we already
+        // voted, it means we already processed it). Since this header got certified, we are sure that all
+        // the data it refers to (ie. its payload and its parents) are available. We can thus continue the
+        // processing of the certificate even if we don't have them in store right now.
+        if !self
+            .processing
+            .get(&certificate.header.round)
+            .map_or_else(|| false, |x| x.contains(&certificate.header.id))
+        {
+            // This function may still throw an error if the storage fails.
+            self.process_header(&certificate.header).await?;
+        }
+
+        // Ensure we have all the ancestors of this certificate yet. If we don't, the synchronizer will gather
```

Comment:
You collect all ancestor certificates? It used to be that having all the certificates of the parents was enough. Do we slow down the DAG to help consensus be faster?

Reply:
It actually doesn't change much. Before we were feeding new certificates to consensus very fast (ie. as soon as we had their parents). But consensus was anyway waiting for a 'last_synchronized_round' from the primary before attempting to order the dag. Now we there is need for this 'last_synchronized_round' as we only feed consensus after we have fully synched the round of the certificates